Append phishfort blocklist

moosthuizen42 opened the pull request:
<!--
Thanks for your contribution!

Please ensure that any applicable requirements below are satisfied before submitting this pull request. This will help ensure a quick and efficient review cycle.
-->

**PhishFort phishing blacklist**

Included PhishFort blacklist in the default blacklist used to perform phishing detection.

**Description**

- ADDED:

  - Fetch of public PhishFort blacklist via cdnjs / GitHub.
  - Inclusion of entries in the PhishFort blacklist into the one used by default (published as part of MetaMask/eth-phishing-detect)
  - Some tests to check for successful detection of entries from the PhishFort blacklist.

**Checklist**

- [x] Tests are included if applicable
- [x] Any added code is fully documented

korn101 commented:
Submitted w/ explanation to @sethkfman via email.

sethkfman commented on src/third-party/PhishingController.ts:
```diff
@@ -135,18 +137,26 @@ export class PhishingController extends BaseController<
+
     if (phishingOpts) {
+      if (pfBlacklist) {
+        phishingOpts.blacklist.push(...pfBlacklist);
```
What if we update the `EthPhishingResponse` interface to contain an optional pfBlacklist []? 

This would not break current implementation and provide a way to identify what list the url originated. The only issue I see with this implementation is that it is not very scalable but I don't know how many other lists we may want to integrate. WDYT? @Gudahtt

korn101 replied:
> 

I like the idea Seth, will wait to hear @Gudahtt's input before making any changes to the PR on our side.

sethkfman replied:
My only other thought to make it a bit more scalable is to use a `externalBlacklists: Dictionary<T>?`. Where the key is the report URL and the value is the []. This way we can append as many blacklists as we want.

owencraston replied:
I second this comment

owencraston replied:
I second this comment

Gudahtt replied:
>This would not break current implementation and provide a way to identify what list the url originated.

Agreed that we need to know where the entry originated. When we tell users that a site is blocked, we provide guidance on how to report incorrectly blocked sites. We need to know where to tell them to go in that situation, we don't want people spamming our repo asking us to unblock sites that are only blocked by phishfort.

One issue with just storing it as a separate list is that it might increase the memory usage of this controller a fair bit. Perhaps we could exclude any entries that are on our main blocklist already, to minimize memory usage.

Gudahtt replied:
Making this scalable to allow other external blocklists sounds like a fine idea to me, but I don't think it needs to be a blocker IMO. We can get this merged first, this one additional list is the more important one. Adding other lists needs a bunch more UI work to be useful in practice, and that hasn't been planned or designed yet, so we have time.

sethkfman replied:
@moosthuizen42 @Gudahtt   What do you think about converting the `phishingOpts.blacklist` to a `phishingOpts.blocklist Dictionary<blockedUrl:urlToReportIncorrectlyBlockedSite>`? 

You would need to add a resolution so duplicates are not to pushed into the dictionary. This dictionary would only contain the keys being a unique set of blockedUrls and values being the corresponding incorrectly blocked report url. If there is a duplicate found between the two lists use the `phishingOpts` report url (MM Report URL: https://github.com/metamask/eth-phishing-detect).

Gudahtt replied:
That might result in even more memory usage than before because of the duplication of the reporting URL.

Or maybe it would get optimized away, who knows. In retrospect, my concern about memory usage was probably premature and I would not block on resolving that. We can measure perf and optimize later.

The critical part is that the lists are separate so we know where to direct users.

moosthuizen42 replied:
Resolved by the PR to `eth-phishing-detect` by @Gudahtt.

Gudahtt commented on src/third-party/PhishingController.ts:
```diff
@@ -135,18 +137,26 @@ export class PhishingController extends BaseController<
 
-    const phishingOpts = await this.queryConfig(this.configUrl);
+    const [phishingOpts, pfBlacklist] = await Promise.all([
+      await this.queryConfig<EthPhishingResponse>(this.configUrlMetaMask),
```
Ideally we'd keep using one list if the other fails, rather than use our fallback. Right now if one failed, both results would get thrown out.

Perhaps `queryConfig` could be updated to catch network errors and log them without throwing?

moosthuizen42 commented:
I have updated the PR with the following:

- Removal of duplicates before appending the PhishFort hotlist.
- Graceful failure if only one list is accessible.
- Updated queryConfig to make the above possible (it no longer throws errors for certain HTTP codes and not for others).

The only thing discussed above but not yet catered for is tying back a specific blocklist entry to a list. The Detector from eth-phishing-detect would likely also need to change to do this, or alternatively the architecture of the PhishingController would need a revamp to allow for multiple PhishingDetector instances.

I would like to get this merged for the sake of getting users protected sooner rather than later, while we work out list separation separately, preferably in its own PR.

moosthuizen42 commented:
@Gudahtt Noted. My understanding is that you intend to make the required changes to `eth-phishing-detect` to allow for multiple list inputs. What are you envisioning in terms of input to the PhishingController? I could suggest the following:

```
{
    version: number;
    tolerance: number;
    blacklist: string[];
    fuzzylist: string[];
    whitelist: string[];
    provider: string; // eg. MetaMask, PhishFort, Umbrella Corporation, etc.
    disputeUrl: string; // Or reportingUrl if that does not cause ambiguity
}[]
```

moosthuizen42 commented:
Thanks @Gudahtt. I'll keep an eye on #7079. Once merged, I intend to update the version of eth-phishing-detect used in our fork, and then integrate both configurations.

Gudahtt commented:
v1.2.0 of `eth-phishing-detect` was just published with support for multiple phishing configurations. Let me know how it works out!

moosthuizen42 commented:
Updated implementation of `PhishingController` to make use of the new input type of `PhishingDetector` as per v1.2 of `eth-phishing-detect`. The new version worked without any issues. State typing and state-checking test also updated to accommodate this.

Please let me know if any other changes are desired, otherwise I think this is ready to be merged.

Gudahtt commented on src/third-party/PhishingController.test.ts:
```diff
@@ -108,6 +112,22 @@ describe('PhishingController', () => {
+  });
+
+  it('should return negative result for unsafe unicode domain if the PhishFort blacklist returns 304', async () => {
+    nock('https://cdn.jsdelivr.net', { allowUnmocked: true })
```
We are actively trying to eliminate network requests from our unit tests, so using `allowUnmocked` is not ideal. I see that this test suite is already mostly unmocked though. I can fix this in a follow-up PR.

moosthuizen42 replied:
Noted. I don't know Jest well, would it be possible to load this from a set of abridged test configs that are part of the repo?

Gudahtt replied:
I'm not _entirely_ sure what you mean, but it sounds like you're asking if we could setup a mock and reuse it on most tests, and the answer to that is definitely yes. Not sure whether Jest could help with that but that's something we can do with nock. We can create a "default" mock and call it in each test, and only mock manually in cases like this where we want to test some special response.

moosthuizen42 replied:
Regarding this, sorry yeah, did mean Jest and/or Nock. Resolving this for now in that case.

Gudahtt commented:
Could you update the `test` method to return the entire result object, instead of just the `result` boolean? We need the config name in order to tell the UI where to direct the user.

moosthuizen42 commented:
Response of `test(...)` updated with newly-defined interface that describes response from the `check(...)` function from `eth-phishing-detect`. All tests that check for boolean results are updated. @Gudahtt, is it necessary to alter the tests to check against the whole response object and not only the boolean property?

Gudahtt commented:
That would be ideal, yeah. We do want to have all properties tested. Though it seems there aren't any fuzzylist tests right now, so those would need to be added. We could use tests for the multi-config setup as well.

Gudahtt commented:
Not sure how much of the test improvements to ask you to complete here. I try to ensure that all new changes are fully tested but not to require adding tests that were already missing in the first place, as a general rule. So if you wanted to add fuzzylist tests, that would be great, but I would accept this with just additional assertions for the `type` property on the current `test(...)` tests, plus a few more tests to ensure that the "allow" and "block" cases work properly with multiple configs, with assertions for the config name.

moosthuizen42 commented:
Done. A sequence of fuzzylist tests added as well. @Gudahtt, once you have taken a look, please confirm that everything you expect to see is there.

![test_results_PhishingController](https://user-images.githubusercontent.com/36568323/165079110-597e9ecb-b4cc-46ba-9d0b-1a850d179840.png)

Other minor changes:
- `EthPhishingDetectResult` was missing `match?: string`, fixed.
- Tests intended to verify behaviour for inaccessible configs changed to HTTP 500 instead of 304.

Gudahtt reviewed (changes requested):
Sorry I'm only just now realizing that the legacy config remains in use until the first update. Could you update the constructor to use the new format as well, passing in the config as an array? I think this would be simpler if we moved directly to the new format, so we wouldn't have to anticipate the two different return types.

moosthuizen42 commented:
> Sorry I'm only just now realizing that the legacy config remains in use until the first update. Could you update the constructor to use the new format as well, passing in the config as an array? I think this would be simpler if we moved directly to the new format, so we wouldn't have to anticipate the two different return types.

Done. Testing references to legacy config have been replaced with default config.

Gudahtt commented on src/third-party/PhishingController.ts:
```diff
@@ -23,6 +23,43 @@ export interface EthPhishingResponse {
+/**
+ * @type EthPhishingDetectResult
+ *
+ * Interface that describes the result of .
+ * @property name - Name of the config on which a match was found. Not returned when using legacy config.
+ * @property version - Version of the config on which a match was found. Not returned when using legacy config.
+ * @property result - Whether a domain was detected as a phishing domain. True means an unsafe domain.
+ * @property match - The matching fuzzylist origin when a fuzzylist match is found. Returned as undefined for non-fuzzy true results.
+ * @property type - The field of the config on which a match was found. 'whitelist' and 'blacklist' can be returned when using legacy config.
+ */
+export interface EthPhishingDetectResult {
+  name?: string; // Not returned for legacy config.
+  version?: string; // Not returned for legacy config.
+  result: boolean;
+  match?: string;  // Returned as undefined for non-fuzzy true results.
+  type: 'all' | 'fuzzy' | 'blocklist' | 'allowlist' | 'whitelist' | 'blacklist';
```
This can be updated now that the legacy config is no longer used. Also the description seemed unfinished so I completed the sentence.

```suggestion
 * Interface that describes the result of the `test` method.
 * @property name - Name of the config on which a match was found.
 * @property version - Version of the config on which a match was found.
 * @property result - Whether a domain was detected as a phishing domain. True means an unsafe domain.
 * @property match - The matching fuzzylist origin when a fuzzylist match is found. Returned as undefined for non-fuzzy true results.
 * @property type - The field of the config on which a match was found.
 */
export interface EthPhishingDetectResult {
  name: string;
  version: string;
  result: boolean;
  match?: string;  // Returned as undefined for non-fuzzy true results.
  type: 'all' | 'fuzzy' | 'blocklist' | 'allowlist';
```

moosthuizen42 replied:
Sorry about the incomplete sentences, there was also a previous mistake of mine that you caught.

Gudahtt reviewed (commented):
LGTM, great work! The tests look fantastic except that the fuzzylist tests don't check the `match` property on the result. But that's fine, I can add that later.

Gudahtt commented:
Looks like I don't have permission to make commits on this branch. I can merge this after my suggestion is addressed or merged, and the lint errors are fixed.

Gudahtt commented on src/third-party/PhishingController.ts:
```diff
@@ -23,6 +23,43 @@ export interface EthPhishingResponse {
+ * @property type - The field of the config on which a match was found.
+ */
+export interface EthPhishingDetectResult {
+  name: string;
+  version: string;
```
Whoops, an earlier suggestion of mine made these mandatory, but they actually aren't present for negative results

```suggestion
  name?: string;
  version?: string;
```

Gudahtt commented:
More lint errors. Please run `yarn lint:fix` locally, and double check that the build and tests pass as well.

moosthuizen42 commented:
@Gudahtt Linting done, build and test complete successfully. Please note that I am only running the `PhishingController.test.ts` tests.

moosthuizen42 commented on src/third-party/PhishingController.ts:
```diff
@@ -135,33 +137,51 @@ export class PhishingController extends BaseController<
 
       default: {
-        throw new Error(
-          `Fetch failed with status '${response.status}' for request '${input}'`,
```
Resolving as okay for now.

Gudahtt reviewed (approved):
LGTM! Holding off a bit on merging this, as I'm going to talk this over with the product teams first and see what it would take to get this change included.

moosthuizen42 commented:
@Gudahtt Any update on this?
